File: homework1/src/warmup.js
function change (amount) {
  if (amount < 0) {
    throw new RangeError("RangeError: amount cannot be negative");
  }
  const result = [];
  let changeRemaining = amount;
  [25,10,5,1].forEach((coin) => {
    result.push(Math.floor(changeRemaining / coin));
    changeRemaining -= coin * Math.floor(changeRemaining / coin);
  });
  return result;
}
 
function stripQuotes (sentence) {
  for (let i = sentence.length; i >= 0; i--) {
    if(sentence.charAt(i) === "\'" || sentence.charAt(i) === "\""){
      sentence = sentence.substring(0,i) + sentence.substring(i+1);
    }
  }
  return sentence;
}
 
function scramble (sentence) {
  var scrambled = "";
  var lengthOfString = sentence.length;
  for (let i = 0; i < lengthOfString; i++) {
    var rand = Math.floor(Math.random() * sentence.length);
    scrambled += sentence.substring(rand,rand+1);
    sentence = sentence.substring(0,rand) + sentence.substring(rand+1);
  }
}
 
function powers (base, max, p) {
  let i = 0;
  while(Math.pow(base, i) <= max) {
    p(Math.pow(base, i));
    i++;
  }
}
 
function cylinder (spec) {
  let {radius, height} = spec;
  radius = (radius !== undefined)?radius:1;
  height = (height !== undefined)?height:1;
  let volume = () => Math.PI * radius * radius * height;
  let surfaceArea = () => 2 * Math.PI * radius * height
    + 2 * Math.PI * radius * radius;
  let widen = (factor) => { radius *= factor };
  let stretch = (factor) => { height *= factor };
  let toString = () => 'Cylinder with radius ${radius} and height ${height}'
  return Object.freeze({
    get radius() {return radius},
    get height() {return height},
    volume, surfaceArea, widen, stretch, toString });
}

const say = function(a){
  let holder;
  if(a){
    holder = a
    return innerScreaming = function(b){
      if(b){
        holder = holder + " " + b
        return innerScreaming;
      }
      else{
        return holder;
      }
    }
  }
  else{
    return '';
  }
}

function* powersGenerator(base, max){
  let power = 0;
  while(max >= Math.pow(base, power)){
    yield Math.pow(base, power);
    power = power+1;
  }
}

function interleave(){
  let args = Array.prototype.slice.call(arguments),
    givenArray = args.shift(),
    weaveArray = [];
  givenArray.forEach(function(eachName, index){
    weaveArray.push(eachName);
    if(args.length > 0){
      weaveArray.push(args.shift());
    }
  })
  if(args.length > 0){
    return weaveArray.concat(args);
  }
  return weaveArray;
}

function makeCryptoFunctions(key, algorithm){
  const crypto = require('crypto');
	
  const encryptFunction = function(secret){
    const cipher = crypto.createCipher(algorithm, key);
	
    let encrypted = cipher.update(secret, 'utf8', 'hex');
    encrypted += cipher.final('hex');
    return encrypted;
  }
  const decryptFunction = function(encrypted){
    const decipher = crypto.createDecipher(algorithm, key);
		
    let decrypted = decipher.update(encrypted, 'hex', 'utf8');
    decrypted += decipher.final('utf8');
    return decrypted;
  }
  return[encryptFunction, decryptFunction];
}

let rp = require('request-promise');

function randomName(spec) {
  let {gender, region} = spec;

  let options = {
    uri: 'http://uinames.com/api/',
    qs: {amount: 1, gender, region},
    json: true,
  };

  return rp(options).then((responseText) => {return responseText.surname+", "+responseText.name});
}

module.exports = {
  change, stripQuotes, scramble, powers, powersGenerator, interleave, randomName, cylinder, say, makeCryptoFunctions
};

File: homework2/warmup.py
import re
import math
import random
import requests
from Crypto.Cipher import AES

def change(amount):
    if amount < 0:
        raise ValueError('amount cannot be negative')
    results = []
    remaining = amount
    for coin in [25, 10, 5, 1]:
        dm = divmod(remaining, coin)
        results.append(dm[0])
        remaining = dm[1]
    return tuple(results)

def strip_quotes(sentence):
    return re.sub(r'\'|\"', '', sentence)

def scramble(sentence):
    letters = list(sentence)
    for item_index in range(0, len(letters)-1):
        random_index = random.randint(item_index, len(letters)-1)
        letters[item_index], letters[random_index] = letters[random_index], letters[item_index]
    return ''.join(letters)

def say(word=''):
    if word == '':
        return word
    def whisper(in_word=''):
        if in_word == '':
            return word
        return say('{} {}'.format(word, in_word))
    return whisper

def triples(max):
    result = []
    for c in range(1, max+1):
        for b in range(1, max+1):
            for a in range(1, max+1):
                if c**2 + b**2 == a**2 and c < b and b < a:
                    result.append((c, b, a))
    return result

def powers(base, maximum):
    value, current_power = 1, 1
    while value <= maximum:
        yield value
        value, current_power = base**current_power, current_power+1

def interleave(first_array, *arguments):
    result = []
    for i in range(max(len(first_array), len(arguments))):
        if len(first_array) > i:
            result.append(first_array[i])
        if len(arguments) > i:
            result.append(arguments[i])
    return result

class Cylinder():
    def __init__(self, radius = 1, height = 1):
        self._radius = radius
        self._height = height

    @property
    def radius(self):
        return self._radius
	
    @property
    def height(self):
        return self._height

    @property
    def volume(self):
        return math.pi * self._height * self._radius ** 2

    @property
    def surface_area(self):
        return 2 * math.pi * self._radius * self._height + 2 * math.pi * self._radius ** 2
	
    def widen(self, factor):
        self._radius *= factor

    def stretch(self, factor):
        self._height *= factor

def make_crypto_functions(key,  iv):
    byte_key, byte_iv = bytes(key, "utf-8"), bytes( iv, "utf-8")
    def encrypt(to_encrypt):
        encrypter = AES.new(byte_key, AES.MODE_CBC, byte_iv)
        return encrypter.encrypt(to_encrypt)

    def decrypt(to_decrypt):
        decrypter = AES.new(byte_key, AES.MODE_CBC, byte_iv)
        return decrypter.decrypt(to_decrypt)
    return (encrypt, decrypt)

def random_name(gender, region):
    r = requests.get("http://uinames.com/api/", params={"gender": gender, "region": region, "amount": 1})
    response = r.text
    if response == '{"error":"Invalid gender"}':
      raise ValueError(response)
    surname = ""
    for i in range(response.find('rname":"') + 8, response.find('","g')):
        surname += response[i]
    name = ""
    for i in range(response.find('"name":"') + 8, response.find('","s')):
        name += response[i]
    return surname + ", " + name